src/Bicep.LangServer/BicepCompilationManager.cs

```diff
@@ -17,72 +21,146 @@ namespace Bicep.LanguageServer
-                
-                // the file is no longer in a state that can be parsed
-                // clear all info to prevent cascading failures elsewhere
-                this.activeContexts.TryRemove(uri, out _);
 
                 // publish a single fatal error diagnostic to tell the user something horrible has occurred
                 // TODO: Tell user how to create an issue on GitHub.
-                this.PublishDocumentDiagnostics(uri, version, new[]
+                var fatalError = new OmniSharp.Extensions.LanguageServer.Protocol.Models.Diagnostic
                 {
-                    new OmniSharp.Extensions.LanguageServer.Protocol.Models.Diagnostic
+                    Range = new Range
                     {
-                        Range = new Range
-                        {
-                            Start = new Position(0, 0),
-                            End = new Position(1, 0),
-                        },
-                        Severity = DiagnosticSeverity.Error,
-                        Message = exception.Message,
-                        Code = new DiagnosticCode("Fatal")
-                    }
-                });
-
-                return null;
+                        Start = new Position(0, 0),
+                        End = new Position(1, 0),
+                    },
+                    Severity = DiagnosticSeverity.Error,
+                    Message = exception.Message,
+                    Code = new DiagnosticCode("Fatal")
+                };
+                
+                // the file is no longer in a state that can be parsed
+                // clear all info to prevent cascading failures elsewhere
+                var closedTrees = CloseCompilationInternal(documentUri, version, fatalError.AsEnumerable());
+
+                return (ImmutableArray<SyntaxTree>.Empty, closedTrees);
             }
         }
 
-        public void CloseCompilation(DocumentUri uri)
+        public void UpsertCompilation(DocumentUri uri, int? version, string text)
         {
-            // remove the active compilation
-            this.activeContexts.TryRemove(uri, out _);
+            var newSyntaxTree = SyntaxTree.Create(uri.GetFileSystemPath(), text);
+            var firstChanges = workspace.UpsertSyntaxTrees(newSyntaxTree.AsEnumerable());
+            var secondChanges = UpdateCompilationInternal(uri, version);
+            
+            var addedTrees = firstChanges.added.Concat(secondChanges.added);
+            var removedTrees = firstChanges.removed.Concat(secondChanges.removed);
 
-            // clear diagnostics for the file
+            foreach (var (entrypointUri, context) in activeContexts)
+            {
+                if (removedTrees.Any(x => context.Compilation.SyntaxTreeGrouping.SyntaxTrees.Contains(x)))
+                {
+                    UpdateCompilationInternal(entrypointUri, null);
+                }
+            }
```

Comment:
This is <= O(activeContexts * removedTrees). May be worth optimizing, but it's not **too** bad for now.

Reply:
wonder what's the best approach to optimize O(n^2), Maybe syntaxTrees becomes a dictionary? worth adding a comment so we don't forget to optimize this path.

Reply:
Yeah I think keeping a cache keyed on SyntaxTree would work just fine.